reference-implementation/src/cipher.rs

```diff
@@ -0,0 +1,253 @@
+use crate::aes_ctr_hmac::AesCtrHmac;
+use crate::header::{Header, KeyId};
+
+use aead::{AeadCore, Key, KeyInit, KeySizeUser, Nonce, Payload};
+use aes::Aes128;
+use aes_gcm::{Aes128Gcm, Aes256Gcm};
+use cipher::consts::{U10, U4, U8};
+use cipher::Unsigned;
+use crypto_common::BlockSizeUser;
+use hkdf::Hkdf;
+use hmac::SimpleHmac;
+use sha2::{Sha256, Sha512};
+
+/// A convenience trait summarizing all of the salient aspects of an AEAD cipher.
+pub trait Aead: aead::Aead + AeadCore + KeyInit + KeySizeUser {}
+impl<T> Aead for T where T: aead::Aead + AeadCore + KeyInit + KeySizeUser {}
+
+/// A convenience trait summarizing all of the salient aspects of a SHA2 digest.
+pub trait Digest: BlockSizeUser + Clone + digest::Digest {}
+impl<T> Digest for T where T: BlockSizeUser + Clone + digest::Digest {}
+
+fn xor_eq(a: &mut [u8], b: &[u8]) {
+    for (b1, b2) in a.iter_mut().zip(b.iter()) {
+        *b1 ^= *b2;
+    }
+}
+
+struct CipherImpl<A: Aead> {
+    /// The label value used as a salt in HKDF
+    pub sframe_label: Vec<u8>,
+
+    /// The `aead_secret` value used as an HKDF PRK
+    pub sframe_secret: Vec<u8>,
+
+    /// The derived AEAD encryption key
+    pub sframe_key: Key<A>,
+
+    /// The derived AEAD encryption salt
+    pub sframe_salt: Nonce<A>,
+    aead: A,
+}
+
+impl<A: Aead> CipherImpl<A> {
+    pub fn new<D: Digest>(kid: KeyId, base_key: &[u8]) -> Self {
+        let mut sframe_label = b"SFrame 1.0 ".to_vec();
+        sframe_label.extend_from_slice(&kid.0.to_be_bytes());
+
+        let (sframe_secret, h) = Hkdf::<D, SimpleHmac<D>>::extract(Some(&base_key), &sframe_label);
+
```

Comment:
Should this be `:extract(Some( &sframe_label), &base_key)`?

Reply:
Yes, thanks.  Wrote to the spec before #127, forgot to change after :)